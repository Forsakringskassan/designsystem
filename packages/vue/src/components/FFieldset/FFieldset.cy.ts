import { defineComponent } from "vue";
import { FFieldsetPageObject } from "../../cypress";
import { FCheckboxField } from "../FCheckboxField";
import { FRadioField } from "../FRadioField";
import FFieldset from "./FFieldset.vue";

const fieldSetRadio = new FFieldsetPageObject(".radio-button-group");
const fieldSetCheckbox = new FFieldsetPageObject(".checkbox-group");

const options = {
    first: ".radio-button:nth(0)",
    second: ".radio-button:nth(1)",
    third: ".radio-button:nth(1)",
    fourth: ".radio-button:nth(1)",
};

const checkBoxOptions = {
    first: ".checkbox:nth(0)",
    second: ".checkbox:nth(1)",
    third: ".checkbox:nth(2)",
    fourth: ".checkbox:nth(3)",
    fifth: ".checkbox:nth(4)",
};

describe("radiobuttons", () => {
    beforeEach(() => {
        const TestComponent = defineComponent({
            components: { FFieldset, FRadioField },
            template: /* HTML */ `
                <f-fieldset horizontal="true">
                    <template #label> Arbetar du deltid? </template>
                    <f-radio-field :value="'test'"> Ja </f-radio-field>
                    <f-radio-field :value="'test2'"> Nej </f-radio-field>
                </f-fieldset>
            `,
        });
        cy.mount(TestComponent);
    });

    it("should mount ffieldset", () => {
        fieldSetRadio.el().should("exist");
    });

    it("should have 2 radiobuttons", () => {
        fieldSetRadio.numberOfRadio().should("equal", 2);
    });

    it("should have labels", () => {
        fieldSetRadio
            .radioButton(options.first)
            .label()
            .should("have.trimmedText", "Ja");
        fieldSetRadio
            .radioButton(options.second)
            .label()
            .should("have.trimmedText", "Nej");
    });

    it("should be able to select first option", () => {
        fieldSetRadio.radioButton(options.first).select();
        fieldSetRadio.radioButton(options.first).isSelected().should("be.true");
        fieldSetRadio
            .radioButton(options.second)
            .isSelected()
            .should("be.false");
    });

    it("should be able to select second option", () => {
        fieldSetRadio.radioButton(options.second).select();
        fieldSetRadio
            .radioButton(options.first)
            .isSelected()
            .should("be.false");
        fieldSetRadio
            .radioButton(options.second)
            .isSelected()
            .should("be.true");
    });
});

describe("chip and horizontal styling", () => {
    it("should contain chip class when chip prop is enabled", () => {
        const TestComponent = defineComponent({
            components: { FFieldset, FRadioField },
            template: /* HTML */ `
                <f-fieldset chip>
                    <template #label> Arbetar du deltid? </template>
                    <f-radio-field :value="'test'"> Ja </f-radio-field>
                    <f-radio-field :value="'test2'"> Nej </f-radio-field>
                </f-fieldset>
            `,
        });
        cy.mount(TestComponent);

        fieldSetRadio.el().should("have.class", "radio-button-group--chip");
    });

    it("should contain horizontal class when horizontal is set to true", () => {
        // Need to have width >= 640px, to be displayed horizontal.
        cy.viewport(1024, 600);
        const TestComponent = defineComponent({
            components: { FFieldset, FRadioField },
            template: /* HTML */ `
                <f-fieldset :horizontal="true">
                    <template #label> Arbetar du deltid? </template>
                    <f-radio-field :value="'test'"> Ja </f-radio-field>
                    <f-radio-field :value="'test2'"> Nej </f-radio-field>
                </f-fieldset>
            `,
        });
        cy.mount(TestComponent);
        fieldSetRadio
            .el()
            .should("have.class", "radio-button-group--horizontal");
    });

    it("should contain horizontal class and chip class when horizontal is set to true and chip prop is enabled", () => {
        // Need to have width >= 640px, to be displayed horizontal.
        cy.viewport(1024, 600);
        const TestComponent = defineComponent({
            components: { FFieldset, FRadioField },
            template: /* HTML */ `
                <f-fieldset horizontal chip>
                    <template #label> Arbetar du deltid? </template>
                    <f-radio-field :value="'test'"> Ja </f-radio-field>
                    <f-radio-field :value="'test2'"> Nej </f-radio-field>
                </f-fieldset>
            `,
        });
        cy.mount(TestComponent);
        fieldSetRadio
            .el()
            .should("have.class", "radio-button-group--horizontal");
        fieldSetRadio.el().should("have.class", "radio-button-group--chip");
    });

    it("should not contain horizontal class or chip class when horizontal is not set to true and chip prop is not enabled", () => {
        const TestComponent = defineComponent({
            components: { FFieldset, FRadioField },
            template: /* HTML */ `
                <f-fieldset>
                    <template #label> Arbetar du deltid? </template>
                    <f-radio-field :value="'test'"> Ja </f-radio-field>
                    <f-radio-field :value="'test2'"> Nej </f-radio-field>
                </f-fieldset>
            `,
        });
        cy.mount(TestComponent);
        fieldSetRadio
            .el()
            .should("not.have.class", "radio-button-group--horizontal");
        fieldSetRadio
            .el()
            .children()
            .should("not.have.class", "radio-button--chip");
    });
});

describe("checkbox", () => {
    beforeEach(() => {
        const TestComponent = defineComponent({
            components: { FFieldset, FCheckboxField },
            template: /* HTML */ `
                <div>
                    <f-fieldset>
                        <template #label>
                            Markera de dagar du har arbetat?
                        </template>
                        <f-checkbox-field value="test1">
                            Måndag
                        </f-checkbox-field>
                        <f-checkbox-field value="test2">
                            Tisdag
                        </f-checkbox-field>
                        <f-checkbox-field value="test3">
                            Onsdag
                        </f-checkbox-field>
                        <f-checkbox-field value="test4">
                            Torsdag
                        </f-checkbox-field>
                        <f-checkbox-field value="test5">
                            Fredag
                        </f-checkbox-field>
                    </f-fieldset>
                </div>
            `,
        });

        cy.mount(TestComponent);
    });

    it("should mount ffieldset", () => {
        fieldSetCheckbox.el().should("exist");
    });

    it("should have 5 radiobuttons", () => {
        fieldSetCheckbox.numberOfCheckboxes().should("equal", 5);
    });

    it("should have labels", () => {
        fieldSetCheckbox
            .checkBox(checkBoxOptions.first)
            .label()
            .should("have.trimmedText", "Måndag");
        fieldSetCheckbox
            .checkBox(checkBoxOptions.second)
            .label()
            .should("have.trimmedText", "Tisdag");
    });

    it("should be able to select and deselect first option", () => {
        fieldSetCheckbox.checkBox(checkBoxOptions.first).select();
        fieldSetCheckbox
            .checkBox(checkBoxOptions.first)
            .isSelected()
            .should("be.true");
        fieldSetCheckbox.checkBox(checkBoxOptions.first).select();
        fieldSetCheckbox
            .checkBox(checkBoxOptions.first)
            .isSelected()
            .should("be.false");
    });

    it("should be able to select and deselect second option", () => {
        fieldSetCheckbox.radioButton(checkBoxOptions.second).select();
        fieldSetCheckbox
            .radioButton(checkBoxOptions.second)
            .isSelected()
            .should("be.true");
        fieldSetCheckbox.radioButton(checkBoxOptions.second).select();
        fieldSetCheckbox
            .radioButton(checkBoxOptions.second)
            .isSelected()
            .should("be.false");
    });

    it("should be able to select and deselect fifth option", () => {
        fieldSetCheckbox.radioButton(checkBoxOptions.fifth).select();
        fieldSetCheckbox
            .radioButton(checkBoxOptions.fifth)
            .isSelected()
            .should("be.true");
        fieldSetCheckbox.radioButton(checkBoxOptions.fifth).select();
        fieldSetCheckbox
            .radioButton(checkBoxOptions.fifth)
            .isSelected()
            .should("be.false");
    });
});
