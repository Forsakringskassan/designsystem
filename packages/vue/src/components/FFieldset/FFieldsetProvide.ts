export const getFieldsetLabelText = Symbol("getFieldsetLabelText");
