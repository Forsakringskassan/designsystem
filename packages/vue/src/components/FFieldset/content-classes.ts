interface LabelOptions {
    hasRadiobutton: boolean;
    hasCheckbox: boolean;
    chip: boolean;
    contentClass: string;
}

/**
 * @internal
 */
export function* contentClasses(options: LabelOptions): Generator<string> {
    const { hasRadiobutton, hasCheckbox, contentClass } = options;

    yield "fieldset__content";

    if (hasRadiobutton) {
        yield "radio-button-group__content";
    }

    if (hasCheckbox) {
        yield "checkbox-group__content";
    }

    yield contentClass;
}
