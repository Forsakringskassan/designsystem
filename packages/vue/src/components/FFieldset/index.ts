export { default as FFieldset } from "./FFieldset.vue";
