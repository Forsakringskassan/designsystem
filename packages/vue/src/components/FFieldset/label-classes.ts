interface LabelOptions {
    hasRadiobutton: boolean;
    hasCheckbox: boolean;
    chip: boolean;
    labelClass: string;
}

/**
 * @internal
 */
export function* labelClasses(options: LabelOptions): Generator<string> {
    const { labelClass } = options;

    yield "fieldset__label";
    yield labelClass;
}
