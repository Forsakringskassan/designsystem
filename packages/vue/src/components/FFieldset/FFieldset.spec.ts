import "html-validate/jest";
import { defineComponent, h } from "vue";
import { VueWrapper, mount, shallowMount } from "@vue/test-utils";
import {
    type ValidityEvent,
    type ValidatableHTMLElement,
    type ValidityMode,
} from "@fkui/logic";
import flushPromises from "flush-promises";
import { ComponentValidityEvent } from "../../types";
import { FIcon } from "../FIcon";
import { getFieldsetLabelText } from "./FFieldsetProvide";
import FFieldset from "./FFieldset.vue";

const defaultSlots = { label: "Label", default: "Content" };

function createWrapper({
    props = {},
    slots = {},
    attrs = {},
    stubs = [],
} = {}): VueWrapper {
    return mount(FFieldset, {
        attrs: { ...attrs },
        props: { id: "someId", name: "someName", ...props },
        slots: { ...slots },
        global: {
            stubs: ["FIcon", ...stubs],
        },
    });
}

function dispatchValidityEvent(
    wrapper: VueWrapper,
    isValid: boolean,
    validityMode: ValidityMode,
    target?: ValidatableHTMLElement,
): void {
    if (!target) {
        target = document.createElement("input");
        target.setAttribute("id", "elementId");
        target.setAttribute("type", "radio");
        wrapper.element.appendChild(target);
    }

    wrapper.element.dispatchEvent(
        new CustomEvent<ValidityEvent>("validity", {
            detail: {
                target,
                elementId: "elementId",
                isValid,
                validityMode,
                validationMessage: "Something went wrong.",
                nativeEvent: "validate",
            },
        }),
    );
}

describe("snapshots", () => {
    it("should match snapshot with generated id attribute", () => {
        const wrapper = createWrapper({ props: { id: undefined } });
        expect(wrapper).toMatchSnapshot();
    });

    it("should match snapshot with label and content", () => {
        const wrapper = createWrapper({ slots: defaultSlots });
        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with label, content and error message", () => {
        const wrapper = createWrapper({
            slots: {
                "error-message": "Something went wrong.",
                ...defaultSlots,
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with label, content and tooltip", () => {
        const wrapper = createWrapper({
            slots: { tooltip: "Tooltip", ...defaultSlots },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it.each`
        validityMode | isValid
        ${"VALID"}   | ${true}
        ${"ERROR"}   | ${false}
        ${"INITIAL"} | ${true}
        ${"INITIAL"} | ${false}
    `(
        "should match snapshot when validityMode is $validityMode",
        async ({ validityMode, isValid }) => {
            const wrapper = createWrapper({
                slots: { label: "Label" },
                attrs: { id: "elementId" },
            });

            dispatchValidityEvent(
                wrapper,
                isValid,
                validityMode,
                wrapper.element as ValidatableHTMLElement,
            );

            await flushPromises();
            wrapper.vm.$forceUpdate();

            expect(wrapper.element).toMatchSnapshot();
        },
    );
});

it("should provide injection for label (legend) text", () => {
    const ChildComponent = defineComponent({
        inject: { getFieldsetLabelText },
        render() {
            return h("div");
        },
    });
    const wrapper = createWrapper({
        slots: {
            label: "Label text",
            default: ChildComponent,
        },
    });

    /* eslint-disable @typescript-eslint/no-explicit-any -- technical debt,
     * should not test methods directly or at least use a proper interface */
    expect(
        (
            wrapper.findComponent(ChildComponent).vm as any
        ).getFieldsetLabelText(),
    ).toBe("Label text");
    /* eslint-enable */
});

describe("attributes", () => {
    it("should pass attributes", () => {
        const wrapper = createWrapper({
            attrs: {
                disabled: true,
                required: true,
                class: "passed-class",
            },
        });
        const fieldset = wrapper.get("fieldset");

        expect(fieldset.attributes("disabled")).toBeDefined();
        expect(fieldset.attributes("required")).toBeDefined();
        expect(fieldset.attributes("class")).toBe("fieldset passed-class");
    });

    it("should only set content class if contentClass prop is set", () => {
        expect.assertions(2);
        const withClass = shallowMount(FFieldset, {
            props: { contentClass: "foo" },
        });
        const withoutClass = shallowMount(FFieldset, {
            props: {},
        });
        expect(withClass.find("div").attributes("class")).toBe(
            "fieldset__content foo",
        );
        expect(withoutClass.find("div").attributes("class")).toBe(
            "fieldset__content",
        );
    });
});

describe("should display error icon when error is present", () => {
    it("without tooltip", async () => {
        expect.assertions(2);
        const wrapper = createWrapper({
            slots: { label: "Label" },
            attrs: { id: "elementId" },
        });
        dispatchValidityEvent(
            wrapper,
            false,
            "ERROR",
            wrapper.element as ValidatableHTMLElement,
        );
        await flushPromises();
        const icon = wrapper.findComponent(FIcon);
        expect(icon.exists()).toBeTruthy();
        expect(icon.attributes("name")).toBe("error");
    });

    it("with tooltip", async () => {
        expect.assertions(2);
        const wrapper = createWrapper({
            slots: { label: "Label", tooltip: "Tooltip" },
            attrs: { id: "elementId" },
        });
        dispatchValidityEvent(
            wrapper,
            false,
            "ERROR",
            wrapper.element as ValidatableHTMLElement,
        );
        await flushPromises();
        const icon = wrapper.findComponent(FIcon);
        expect(icon.exists()).toBeTruthy();
        expect(icon.attributes("name")).toBe("error");
    });
});

describe("events", () => {
    it("should pass listeners", async () => {
        const foobar = jest.fn();
        const wrapper = createWrapper({
            attrs: { onFoobar: foobar },
        });
        const element = wrapper.get("fieldset");
        await element.trigger("foobar");
        expect(foobar).toHaveBeenCalled();
    });
});

describe("onValidity should only handle events from itself", () => {
    it("handles events if event.target is itself", async () => {
        const wrapper = createWrapper({
            slots: { label: "Label" },
            attrs: { id: "elementId" },
        });
        const onComponentValidityListener = jest.fn();
        wrapper.element.addEventListener(
            "component-validity",
            onComponentValidityListener,
        );

        dispatchValidityEvent(
            wrapper,
            false,
            "INITIAL",
            wrapper.element as ValidatableHTMLElement,
        );

        await flushPromises();
        wrapper.vm.$forceUpdate();

        expect(onComponentValidityListener).toHaveBeenCalled();
    });

    it.each`
        element       | inputType     | description
        ${"input"}    | ${"radio"}    | ${"should not"}
        ${"input"}    | ${"checkbox"} | ${"should not"}
        ${"input"}    | ${"text"}     | ${"should not"}
        ${"textarea"} | ${undefined}  | ${"should not"}
        ${"select"}   | ${undefined}  | ${"should not"}
    `(
        "$description dispatch component-validity event if element is $element and type is $inputType",
        ({ element, inputType }) => {
            const wrapper = createWrapper({
                slots: { label: "Label" },
                attrs: { id: "elementId" },
            });
            const onComponentValidityListener = jest.fn();
            wrapper.element.addEventListener(
                "component-validity",
                onComponentValidityListener,
            );
            const inputElement = document.createElement(element);
            inputElement.setAttribute("id", "elementId");
            inputElement.setAttribute("type", inputType);
            wrapper.element.appendChild(inputElement);

            dispatchValidityEvent(wrapper, false, "INITIAL", inputElement);
            expect(onComponentValidityListener).not.toHaveBeenCalled();
        },
    );

    it.each`
        inputType
        ${"radio"}
        ${"checkbox"}
    `(
        "should not dispatch component-validity event if input $inputType belongs to nestled fieldset",
        ({ inputType }) => {
            const wrapper = createWrapper({
                slots: { label: "Label" },
                attrs: { id: "elementId" },
            });
            const onComponentValidityListener = jest.fn();
            wrapper.element.addEventListener(
                "component-validity",
                onComponentValidityListener,
            );

            const fieldset = document.createElement("fieldset");
            wrapper.element.appendChild(fieldset);

            const inputElement = document.createElement("input");
            inputElement.setAttribute("id", "elementId");
            inputElement.setAttribute("type", inputType);

            fieldset.appendChild(inputElement);

            dispatchValidityEvent(wrapper, false, "INITIAL", inputElement);
            expect(onComponentValidityListener).not.toHaveBeenCalled();
        },
    );
});

describe("onValidity should set focusElementId in ComponentValidityEvent", () => {
    it("should set focusElementId to own id if having no child input element", async () => {
        const wrapper = createWrapper({
            slots: { label: "Label" },
            props: { id: "elementId" },
        });
        const onComponentValidityListener = jest.fn();
        wrapper.element.addEventListener(
            "component-validity",
            onComponentValidityListener,
        );

        dispatchValidityEvent(
            wrapper,
            false,
            "INITIAL",
            wrapper.element as ValidatableHTMLElement,
        );

        await flushPromises();
        wrapper.vm.$forceUpdate();

        const componentValidityEvent = onComponentValidityListener.mock
            .calls[0][0].detail as ComponentValidityEvent;
        expect(componentValidityEvent.elementId).toBe("elementId");
        expect(componentValidityEvent.focusElementId).toBe("elementId");
    });

    it("should set focusElementId to first child input element", async () => {
        const wrapper = createWrapper({
            slots: { label: "Label" },
            props: { id: "elementId" },
        });
        const onComponentValidityListener = jest.fn();
        wrapper.element.addEventListener(
            "component-validity",
            onComponentValidityListener,
        );

        const inputElement1 = document.createElement("input");
        inputElement1.setAttribute("id", "child1Id");
        wrapper.element.appendChild(inputElement1);

        const inputElement2 = document.createElement("input");
        inputElement2.setAttribute("id", "child2Id");
        wrapper.element.appendChild(inputElement2);

        dispatchValidityEvent(
            wrapper,
            false,
            "INITIAL",
            wrapper.element as ValidatableHTMLElement,
        );

        await flushPromises();
        wrapper.vm.$forceUpdate();

        const componentValidityEvent = onComponentValidityListener.mock
            .calls[0][0].detail as ComponentValidityEvent;
        expect(componentValidityEvent.elementId).toBe("elementId");
        expect(componentValidityEvent.focusElementId).toBe("child1Id");
    });
});

describe("html-validate", () => {
    describe("content model", () => {
        it("should be allowed where flow is expected", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <div>
                    <f-fieldset name="foo">
                        <template #label> Label </template>
                    </f-fieldset>
                </div>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not be allowed where phrasing is expected", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <span>
                    <f-fieldset name="foo">
                        <template #label> Label </template>
                    </f-fieldset>
                </span>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });
    });

    describe("props", () => {
        it("should allow non-empty name", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset name="foo">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow empty name", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset name>
                    <template #label> Label </template>
                </f-fieldset>
                <f-fieldset name="">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: Attribute "name" is missing value (attribute-allowed-values) at inline:2:29:
                  1 |
                > 2 |                 <f-fieldset name>
                    |                             ^^^^
                  3 |                     <template #label> Label </template>
                  4 |                 </f-fieldset>
                  5 |                 <f-fieldset name="">
                Selector: f-fieldset:nth-child(1)
                error: Attribute "name" has invalid value "" (attribute-allowed-values) at inline:5:29:
                  3 |                     <template #label> Label </template>
                  4 |                 </f-fieldset>
                > 5 |                 <f-fieldset name="">
                    |                             ^^^^
                  6 |                     <template #label> Label </template>
                  7 |                 </f-fieldset>
                  8 |
                Selector: f-fieldset:nth-child(2)"
            `);
        });

        it("should allow content class", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset content-class="foo">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow multiple content classes", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset content-class="foo bar">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: Attribute "content-class" has invalid value "foo bar" (attribute-allowed-values) at inline:2:44:
                  1 |
                > 2 |                 <f-fieldset content-class="foo bar">
                    |                                            ^^^^^^^
                  3 |                     <template #label> Label </template>
                  4 |                 </f-fieldset>
                  5 |
                Selector: f-fieldset"
            `);
        });

        it("should allow label class", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset label-class="foo">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow multiple label classes", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset label-class="foo bar">
                    <template #label> Label </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: Attribute "label-class" has invalid value "foo bar" (attribute-allowed-values) at inline:2:42:
                  1 |
                > 2 |                 <f-fieldset label-class="foo bar">
                    |                                          ^^^^^^^
                  3 |                     <template #label> Label </template>
                  4 |                 </f-fieldset>
                  5 |
                Selector: f-fieldset"
            `);
        });
    });

    describe("slots", () => {
        it("should require label slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ ` <f-fieldset> </f-fieldset> `;
            expect(markup).toMatchInlineCodeframe(`
                "error: <f-fieldset> component requires slot "label" to be implemented (vue/required-slots) at inline:1:3:
                > 1 |  <f-fieldset> </f-fieldset>
                    |   ^^^^^^^^^^
                Selector: f-fieldset"
            `);
        });

        it("should allow phrasing in label slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label>
                        <span> Label </span>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should allow headings in label slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label>
                        <h2>Label</h2>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow block elements in label slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label>
                        <div>Label</div>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: <div> element is not permitted as content under slot "label" (<f-fieldset>) (element-permitted-content) at inline:4:26:
                  2 |                 <f-fieldset>
                  3 |                     <template #label>
                > 4 |                         <div>Label</div>
                    |                          ^^^
                  5 |                     </template>
                  6 |                 </f-fieldset>
                  7 |
                Selector: f-fieldset > template > div"
            `);
        });

        it("should allow phrasing in description slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label> Label </template>
                    <template #description>
                        <span> Lorem ipsum </span>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow block elements in description slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label> Label </template>
                    <template #description>
                        <div>Lorem ipsum</div>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: <div> element is not permitted as content under slot "description" (<f-fieldset>) (element-permitted-content) at inline:5:26:
                  3 |                     <template #label> Label </template>
                  4 |                     <template #description>
                > 5 |                         <div>Lorem ipsum</div>
                    |                          ^^^
                  6 |                     </template>
                  7 |                 </f-fieldset>
                  8 |
                Selector: f-fieldset > template:nth-child(2) > div"
            `);
        });

        it("should allow f-tooltip in tooltip slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label> Label </template>
                    <template #tooltip>
                        <f-tooltip screen-reader-text="lorem ipsum">
                            Lorem ipsum
                        </f-tooltip>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`""`);
        });

        it("should not allow arbitrary elements in tooltip slot", () => {
            expect.assertions(1);
            const markup = /* HTML */ `
                <f-fieldset>
                    <template #label> Label </template>
                    <template #tooltip>
                        <div>Lorem</div>
                        <span>ipsum</span>
                    </template>
                </f-fieldset>
            `;
            expect(markup).toMatchInlineCodeframe(`
                "error: <div> element is not permitted as content under slot "tooltip" (<f-fieldset>) (element-permitted-content) at inline:5:26:
                  3 |                     <template #label> Label </template>
                  4 |                     <template #tooltip>
                > 5 |                         <div>Lorem</div>
                    |                          ^^^
                  6 |                         <span>ipsum</span>
                  7 |                     </template>
                  8 |                 </f-fieldset>
                Selector: f-fieldset > template:nth-child(2) > div
                error: <span> element is not permitted as content under slot "tooltip" (<f-fieldset>) (element-permitted-content) at inline:6:26:
                  4 |                     <template #tooltip>
                  5 |                         <div>Lorem</div>
                > 6 |                         <span>ipsum</span>
                    |                          ^^^^
                  7 |                     </template>
                  8 |                 </f-fieldset>
                  9 |
                Selector: f-fieldset > template:nth-child(2) > span"
            `);
        });
    });

    describe("screen reader text", () => {
        it("one child should only have checkbox checked screen reader text", async () => {
            const TestComponent = defineComponent({
                components: { FFieldset },
                template: /* HTML */ `
                    <f-fieldset>
                        <input type="checkbox" />
                    </f-fieldset>
                `,
            });
            const wrapper = mount(TestComponent);
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            const element = wrapper.get("[data-test='checked-boxes']");
            expect(element.text()).toBe("Kryssruta ej kryssad");
        });

        it("other than one child should have numbers of and checked children screen reader text", async () => {
            const TestComponent = defineComponent({
                components: { FFieldset },
                template: /* HTML */ `
                    <f-fieldset>
                        <input type="checkbox" />
                        <input type="checkbox" />
                    </f-fieldset>
                `,
            });
            const wrapper = mount(TestComponent);
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            const content = wrapper.get("[data-test='checked-boxes']");
            const label = wrapper.get(".sr-only");
            expect(content.text()).toBe("0 kryssad av 2");
            expect(label.text()).toBe("Grupp med 2 kryssrutor");
        });

        it("no child should have no screen reader text", async () => {
            const TestComponent = defineComponent({
                components: { FFieldset },
                template: /* HTML */ ` <f-fieldset> </f-fieldset> `,
            });
            const wrapper = mount(TestComponent);
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            await wrapper.vm.$nextTick();
            const content = wrapper.find("[data-test='checked-boxes']");
            const label = wrapper.find(".sr-only");
            expect(content.exists()).toBe(false);
            expect(label.exists()).toBe(false);
        });
    });
});
