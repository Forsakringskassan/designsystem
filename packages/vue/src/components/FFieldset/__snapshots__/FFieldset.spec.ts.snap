// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots should match snapshot when validityMode is ERROR 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <span
      class="label__message label__message--error"
    >
      <f-icon-stub
        class="label__icon--left"
        library="f"
        name="error"
      />
       Something went wrong.
    </span>
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot when validityMode is INITIAL 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <!--v-if-->
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot when validityMode is INITIAL 2`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <!--v-if-->
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot when validityMode is VALID 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <!--v-if-->
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot with generated id attribute 1`] = `
<fieldset id="fkui-vue-element-0001" class="fieldset">
  <legend class="label fieldset__label">
    <!--v-if-->
    <!--v-if-->
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div class="fieldset__content"></div>
</fieldset>
`;

exports[`snapshots should match snapshot with label and content 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <!--v-if-->
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    Content
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot with label, content and error message 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label fieldset__label"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    Something went wrong.
    
  </legend>
  <!--v-if-->
  <!--v-if-->
  <div
    class="fieldset__content"
  >
    
    Content
    
  </div>
</fieldset>
`;

exports[`snapshots should match snapshot with label, content and tooltip 1`] = `
<fieldset
  class="fieldset"
  id="someId"
>
  <legend
    class="label sr-only"
  >
    
    Label
    
    <!--v-if-->
    
    
    
    <!--v-if-->
    
  </legend>
  <!--v-if-->
  
  <div
    class="sr-separator"
  >
    <div
      aria-hidden="true"
      class="tooltip-before"
    >
      <div
        class="label tooltip-before__label"
      >
        
        Label
        
      </div>
    </div>
    
    Tooltip
    
  </div>
  <!--v-if-->
  
  <div
    class="fieldset__content"
  >
    
    Content
    
  </div>
</fieldset>
`;
